CSM-337 Extrapolate timestamps in case of slot duration change.

ksaric opened the pull request:
https://issues.serokell.io/issue/CSM-337

Most of the work was done by @ajnsit - https://github.com/input-output-hk/cardano-sl/pull/1107
There are some things I need to fix (imports), comments, but this should be it.

I found this code kind of confusing, it seems to be scattered everywhere.

@gromakovsky Can you take a look at this? The most confusing is the fact we track two epoch - the last one and the penult one. I don't even know why, this might be wrong, depending on how we track that - https://github.com/input-output-hk/cardano-sl/compare/ksaric/CSM-337?expand=1#diff-9750ad2f43ce45f68873f4b73aa3c651R87

gromakovsky commented:
I am going to review it and I can already say that I think commit message (as well issue title) is outdated, because instead of extrapolation we decided to store full slotting data.

gromakovsky commented:
> The most confusing is the fact we track two epoch - the last one and the penult one.

First, I need to clarify one thing which is not obvious: _the last epoch_ means _the last epoch for which we know `EpochSlottingData`_ and _the penult epoch_ means _the penult epoch for which we know `EpochSlottingData`_. So _the last_ epoch is not the _current_ epoch. Actually, we store `EpochSlottingData` for the current epoch (called _penult_) and for the next epoch (called _last_).
The reason we store slotting data for the current epoch is quite clear: it's needed to be able to know the current slot. The reason we store slotting data for the next epoch is that sometimes we need to know when the next slot starts and it can in another epoch.
The reason we don't store slotting data for future slots is that we don't know it (it can be changed).
The reason we don't store slotting data for older slots is that we don't need it and want to save some space. This sentence is no longer true and this PR is dedicated to changing it (so that we will store slotting data for all epochs).

Maybe I am too verbose, but I hope it answers this particular question.

Another question is whether we still need to store slotting data for these two epochs. Conceptually we know that we always know slotting data for at least two epochs. Whether we need to make it explicit or not is mostly a convenience concern. I think it should be more convenient to make it explicit that we have at least 1 `EpochSlottingData` (in this case `getLastEpochIndex` doesn't have to return `Maybe`), but probably it's won't be beneficial to store more than 1.

_probably I had to think about it earlier_

gromakovsky commented on infra/Pos/Arbitrary/Slotting.hs:
```diff
@@ -6,13 +6,12 @@ import           Test.QuickCheck                   (Arbitrary (..))
 instance Arbitrary SlottingData where
     arbitrary = genericArbitrary
-    shrink = genericShrink
+    shrink = genericShrink    
```
It's recommended to setup your editor to remove trailing whitespaces automatically.

ksaric replied:
Added.

gromakovsky commented on infra/Pos/Slotting/Impl/Ntp.hs:
```diff
@@ -152,8 +151,9 @@ data SlotStatus
 ntpGetCurrentSlot
-    :: NtpMode m
-    => NtpSlottingVar -> m (Maybe SlotId)
+    :: (NtpMode m, MonadThrow m)
```
I think it would be better to include `MonadThrow` into `NtpMode`.

gromakovsky commented on infra/Pos/Slotting/Impl/Ntp.hs:
```diff
@@ -168,11 +168,12 @@ ntpGetCurrentSlot var = ntpGetCurrentSlotImpl var >>= \case
     printSlottingData = do
-        sd <- getSlottingData
+        eli <- getEpochLastIndex
+        sd  <- getEpochSlottingData eli
```
There should be `getLastEpochSlottingData` which returns `EpochSlottingData` instead of `Maybe EpochSlottingData` and it should be user here.

gromakovsky commented on infra/Pos/Slotting/Impl/Ntp.hs:
```diff
@@ -168,11 +168,12 @@ ntpGetCurrentSlot var = ntpGetCurrentSlotImpl var >>= \case
-        sd <- getSlottingData
+        eli <- getEpochLastIndex
+        sd  <- getEpochSlottingData eli
         logWarning $ "Slotting data: " <> show sd
```
I would change it to `Last epoch slotting data:`

gromakovsky commented on infra/Pos/Slotting/Impl/Ntp.hs:
```diff
@@ -184,19 +185,19 @@ ntpGetCurrentSlotInaccurate var = do
+          let setLastSlot s = atomically $ STM.modifyTVar' var (nssLastSlot %~ max s)
           whenJust res setLastSlot
-          pure $ maybe (OutdatedSlottingData penult) CurrentSlot res
+          pure $ maybe (OutdatedSlottingData lastIndex) CurrentSlot res
```
I see you changed the semantics from attaching penult index to last index, but didn't change a comment to `OutdatedSlottingData` and the place where it's printed.

gromakovsky commented on infra/Pos/Slotting/Impl/Simple.hs:
```diff
@@ -47,21 +46,30 @@ mkSimpleSlottingVar = atomically $ newTVar $ SimpleSlottingState $ unflattenSlot
 
-getCurrentSlotSimple :: SimpleSlottingMode m => SimpleSlottingVar -> m (Maybe SlotId)
+getCurrentSlotSimple 
+    :: (SimpleSlottingMode m, MonadThrow m)
```
Here I also think it would be better to add `MonadThrow` to `SimpleSlottingMode`.

gromakovsky commented on infra/Pos/Slotting/Impl/Simple.hs:
```diff
@@ -47,21 +46,30 @@ mkSimpleSlottingVar = atomically $ newTVar $ SimpleSlottingState $ unflattenSlot
+    -> m SlotId
 getCurrentSlotBlockingSimple var = do
-    penult <- sdPenultEpoch <$> getSlottingData
+    penult <- getEpochLastIndex
```
Hm, but it's not `penult` anymore :/

gromakovsky replied:
Though the logic looks correct, you only need to change the name.

gromakovsky commented on infra/Pos/Slotting/Impl/Sum.hs:
```diff
@@ -47,25 +47,33 @@ askSlottingContextSum = view (lensOf @SlottingContextSum)
 
-getCurrentSlotSum :: SlotsSumEnv ctx m => m (Maybe SlotId)
+getCurrentSlotSum 
+    :: (SlotsSumEnv ctx m, MonadThrow m)
```
Again, it's better to add `MonadThrow` to `SlotsSumEnv`.

gromakovsky commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -32,21 +37,28 @@ approxSlotUsingOutdated penult t = do
+    let penultRes = computeSlotUsingEpoch systemStart approxCurTime sdPenultEpoch sdPenult
+    let lastRes   = computeSlotUsingEpoch systemStart approxCurTime (succ lastEpochI) sdLast
+
     return $ penultRes <|> lastRes
```
Hm, this code should be reworked. Currently it checks only the last two epochs. So it will return `Nothing` for an old epoch. But the task is to improve it! It should find an appropriate epoch and use its slotting data.
If I recall correctly, there was right logic in the original PR.
It should return `Nothing` only for timestamps which are after the last epoch from `SlottingData`.

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,38 @@ import           Control.Monad.Trans (MonadTrans)
+    getEpochLastIndex :: m EpochIndex
 
-    waitPenultEpochEquals :: EpochIndex -> m ()
+    -- TODO(ks): Maybe add other usefull functions directly?
```
Well, ideally this class should be deleted and should become a simple reader over `SlottingVar` instead. But that's for future I think.

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,38 @@ import           Control.Monad.Trans (MonadTrans)
 
-    waitPenultEpochEquals :: EpochIndex -> m ()
+    -- TODO(ks): Maybe add other usefull functions directly?
+    -- getEpochPenultIndex :: m EpochIndex
```
I think this is not needed as I wrote in a standalone comment to this PR (i. e. we should completely deprecated everything `penult` :)).

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,38 @@ import           Control.Monad.Trans (MonadTrans)
-    waitPenultEpochEquals :: EpochIndex -> m ()
+    -- TODO(ks): Maybe add other usefull functions directly?
+    -- getEpochPenultIndex :: m EpochIndex
+    -- getLastEpochSlottingData :: m (Maybe EpochSlottingData)
```
This is good, but it should return `m EpochSlottingData` instead.

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,38 @@ import           Control.Monad.Trans (MonadTrans)
+    -- TODO(ks): Maybe add other usefull functions directly?
+    -- getEpochPenultIndex :: m EpochIndex
+    -- getLastEpochSlottingData :: m (Maybe EpochSlottingData)
+    -- getPenultEpochSlottingData :: m (Maybe EpochSlottingData)
```
This is not needed, because see above (about `penult`).

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,38 @@ import           Control.Monad.Trans (MonadTrans)
-    putSlottingData :: SlottingData -> m ()
+    putEpochSlottingData :: EpochIndex -> EpochSlottingData -> m ()
+
+    waitPenultEpochEquals :: EpochIndex -> m ()
```
I would rename it to `waitLastEpochEquals`.

gromakovsky commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +37,96 @@ data EpochSlottingData = EpochSlottingData
+    :: PenultSlottingData 
+    -> EpochSlottingData 
+    -> SlottingData
+createInitSlottingData psd esd = SlottingData validInitialSlottingData
```
Again, I think we don't need penult epoch slotting data, as I wrote earlier. Probably I had to think earlier, sorry. It doesn't seem to give any benefit anymore and adds extra complication.
So previously we had effectively a map with exactly two elements and now we are going to have a map with at least 1 element.

gromakovsky replied:
So I think this safe constructor should have type `(EpochIndex, EpochSlottingData) -> Map EpochIndex EpochSlottingData -> SlottingData`.

gromakovsky commented on infra/Pos/Slotting/Util.hs:
```diff
@@ -90,9 +72,30 @@ getSlotStartEmpatically
+    getEpochSlottingData penultEpoch >>= maybeThrow (SEUnknownLastEpoch penultEpoch)
+
+-- | Get last epoch empatically, which means
+-- that function throws exception when last epoch is unknown.
```
This function doesn't make sense, because the last epoch slotting data is always known.

gromakovsky commented on src/Pos/Block/Slog/Logic.hs:
```diff
@@ -290,8 +291,12 @@ slogRollbackBlocks blunds = do
     slogPutLastSlots newLastSlots
-    putSlottingData =<< GS.getSlottingData
+    slotData <- M.toList . getSlottingDataMap <$> GS.getSlottingData
+    forM_ slotData (uncurry putEpochSlottingData)
```
It can be optimized, but maybe in another task.

ksaric commented:
![epoch slotting data](https://user-images.githubusercontent.com/6264437/28613560-fc911ee2-71f2-11e7-8881-f0067768366a.jpg)

I'm not sure how having all `EpochSlottingData` (having it for all `Epoch`s) changes the fact that we have `penult` and `last` epochs.
My understanding is that we still conceptually have these two, we just have others as well :up: 

> The reason we store slotting data for the current epoch is quite clear: it's needed to be able to know the current slot. The reason we store slotting data for the next epoch is that sometimes we need to know when the next slot starts and it can in another epoch.

This cleared up some things, thanks.

But it would be more understandable (to me, at least) to rename:
- `penult` to `current`
- `last` to [`next`, `following`, ...]

gromakovsky commented:
> I'm not sure how having all EpochSlottingData (having it for all Epochs) changes the fact that we have penult and last epochs.

It doesn't change this fact, we still have at least two `EpochSlottingData`s. But I don't see any place where we need to rely on having at least two `EpochSlottingData`s. However, I see some places where we rely on non-emptiness of `SlottingData`.
And making `penult` epoch slotting data somewhat _special_ complicates things. So I propose to make only last epoch slotting data _special_ (and maybe write somewhere that it's the next epoch after the current one).

ksaric commented:
> But I don't see any place where we need to rely on having at least two `EpochSlottingData`s

But you wrote:

> The reason we store slotting data for the next epoch is that sometimes we need to know when the next slot starts and it can in another epoch.

So we need the next slot information. And we need current slot information. So that makes both of them required? And that's two?

> And making penult epoch slotting data somewhat special complicates things. So I propose to make only last epoch slotting data special (and maybe write somewhere that it's the next epoch after the current one).

:confused: 

There is no emoji for "head explodes", so I had to stick with "confused". Not sure what you mean by "special"? Do you mean we only need _last epoch_ `EpochSlottingData` and don't need the _current epoch_ `EpochSlottingData`?

gromakovsky commented:
Ok, I will try to explain slightly differently.

Previously we stored exactly two epoch slotting data. There is no data type representing a map with size = 2, so we it was implemented by defining ad-hoc data type with 2 `EpochSlottingData`s and one `EpochIndex`. We could also implement it as `Map EpochIndex EpochSlottingData`, it wouldn't be much different. It's a trade-off between safety of operations and extra amount of code.

In this PR we want to store all epoch slotting data from 0 to _next_, where _next_ must be current epoch + 1. We have various options:
1. Store it as a plain map.
2. Store it as a custom type with invariant that it has at least 1 item (there are different ways to represent it, but the main idea is that this invariant belongs to this type.
3. Same as (2) but at least 2 items.

Looking at the code, I see advantage of (2) over (1), because in some places you really need to have `getLastEpochSlottingData :: m EpochSlottingData` (w/o maybe). So it's move convenient to make it invariant of `SlottingData` rather than check it in places where it is used. However, I don't see places where it is necessary to access penult epoch. Maybe they exist in the current code, but they can be eliminated. I didn't dig very deep, can't say for sure.

Conclusions:
1. Previously we stored penult epoch slotting data explicitly because we wanted to store **exactly** two (not more and not less) epoch slotting datas, not because we wanted to rely on its presence (maybe I am forgetting something though, but I believe it's true).
2. This is all about trade-offs between safety and amount of code and in general there is no ideal solution. For example, `getCurrentSlot` still returns `Maybe` even if we know that `EpochSlottingData` for the current epoch exists in `SlottingData`. And in some places we implicitly rely on this fact (we know it, but we don't guarantee it in code and we have to use `getSlotStartEmpatically` which can throw an exception).
3. If you see an advantage in statically ensuring that `SlottingData` contains at least two items (not 1), I don't mind. But if there is no advantage, I think it's better not to do it, it will make code simpler.

ksaric commented:
@gromakovsky Ok, so I fixed some things. 

1. Your review points
2. I fixed the names which were confusing to me:
- `penult` -> `current`
- `last` -> `next`

I added the `M` suffix to the typeclass functions, seems to be more informative and doesn't conflict with the functions from `Util` -https://github.com/input-output-hk/cardano-sl/pull/1217/files#diff-2bc1b266461390adfcf818e8e64cfa11R17

https://github.com/input-output-hk/cardano-sl/pull/1217/files#diff-3aacd87b5c16bb262833bd963cafd4aeR221

gromakovsky commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
+    => EpochIndex
+    -> Timestamp
+    -> m SlotId
 approxSlotUsingOutdated penult t = do
```
This function still uses `penult` as the name of argument. And there are more usages of this word in your branch. I think it makes sense to remove all of them because you changed the naming:

```
gromak ~/Work/cardano-sl (ksaric/CSM-337|…) % scripts/grep.sh -i "penult"                                                                                                                                                                                                 0 :)
test/Test/Pos/Block/Logic/Mode.hs:                                                 waitPenultEpochEqualsDefault)
test/Test/Pos/Block/Logic/Mode.hs:    waitPenultEpochEquals = waitPenultEpochEqualsDefault
test/Test/Pos/Block/Logic/Mode.hs:    waitPenultEpochEquals = waitPenultEpochEqualsDefault
update/Pos/Update/Poll/Logic/Base.hs:            ("can't update slotting data, stored penult epoch is "%int%
update/Pos/Update/DB.hs:    esdPenult :: EpochSlottingData
update/Pos/Update/DB.hs:    esdPenult = EpochSlottingData
update/Pos/Update/DB.hs:    genesisSlottingData = createInitSlottingData esdPenult esdLast
infra/Pos/Slotting/MemState/Holder.hs:        penultEpoch <- getCurrentEpochIndex <$> readTVar var
infra/Pos/Slotting/MemState/Holder.hs:        when (penultEpoch /= target) retry
infra/Pos/Slotting/Impl/Util.hs:approxSlotUsingOutdated penult t = do
infra/Pos/Slotting/Impl/Util.hs:        if | t < epochStart -> SlotId (penult + 1) minBound
infra/Pos/Slotting/Impl/Util.hs:           | otherwise      -> outdatedEpoch systemStart t (penult + 1) sdLast
lwallet/Pos/Wallet/Light/Mode.hs:    waitPenultEpochEquals = error "notImplemented"
```

gromakovsky replied:
And also instead of passing penult (current) here it makes more sense to pass last (next). Do you agree? I don't insist on fixing it, but seems to be quite trivial.

gromakovsky replied:
Hm, it took me some time understand what's going on here and why we pass `penult`. I think we shouldn't pass `penult` at all, because it can be fetched from `MonadSlotsData`. Moreover, I find it dangerous to pass it as argument, because `TVar` with slotting data can change while this function is executed and we may get inconsistent result. I think we should get next epoch and its slotting data in one function call.

ksaric replied:
I will rename other `penult` names, you are correct. I will remove the argument and call it directly, noted.

ksaric replied:
Also, seems we don't need any parameters to `OutdatedSlottingData` but I'll leave it as it is. I think you guys probably know more about this.

gromakovsky commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
         unflattenSlotId $
         flattenEpochIndex epoch + fromIntegral ((curTime - start) `div` duration)
 
+-- TODO(ks): Move to MonadSlotsData?
```
No, we want to get rid of `MonadSlotsData` because it's useless abstraction (it doesn't have multiple implementations, doesn't really hide anything, etc.). Sooner or later it will be replaced by `(MonadReader ctx m, HasSlottingVar ctx)`.

ksaric replied:
Is it really that bad? I mean, I know I'm missing quite a lot of knowledge regarding Haskell, but it doesn't seem that useless.
Seems to me we could hide _all the module functions_ behind the `MonadSlotsData` typeclass. That would enable us to export only the typeclass and hide all other implementation detail in the module itself - it would enable us to refactor and change code more easily.
Doing that _we wouldn't have any additional dependencies on the module other than the typeclass_, which also supports the ideas like having a single responsibility, easier testing, easier refactoring, modularity...
I'll remove the comment, but still, seems kind of interesting (to me).

gromakovsky commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
+    -- We partially apply the function and then iterate over all epoch indexes.
+    mSlotId         <- forM allEpochIndex (findSlot systemStart approxCurTime)
+
+    pure $ head $ catMaybes mSlotId
```
I would use `pure $ foldr (<|>) Nothing mSlotId` instead. It avoids partial function and should be more efficient.

gromakovsky replied:
Well, to be honest, I am afraid that it can be quite inefficient. Especially because you start from the oldest epochs. This function is called quite often and the size of `SlottingData` can will become bigger and bigger.
And this code will traverse all epochs from the 0-th to the _next_ one and do some computations. I propose to do two modifications:
1. Start from the newest epoch.
2. Stop if the result is computed successfully (so it should be `foldM` most likely).

ksaric replied:
You are right. Honestly, I didn't even consider performance. Will change.

ksaric replied:
I wrote a small function that should be efficient. I didn't figure out how `foldM` was supposed to fit in this scenario. 
I was checking out other options, I guess I could have used `MonadLoops` with `ListT` - there is a function `untilJust` there. But `MonadLoops` is missing as a dependency.

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
+
+    getCurrentEpochSlottingDataM = lift getCurrentEpochSlottingDataM
+
+    getNextEpochSlottingDataM = lift getCurrentEpochSlottingDataM
```
It's a bug I guess.

ksaric replied:
Yes :man_facepalming:

gromakovsky commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
+
+    getEpochSlottingDataM = lift . getEpochSlottingDataM
+
+    putEpochSlottingDataM = (lift .) . putEpochSlottingDataM
```
Btw, there is `...` in `Universum` which is a var-arg generalization of `.`. So you can right `lift ... f` regadless of the number of arguments `f` has.

ksaric replied:
I remember that but forgot it's purpose. Thanks, I'll use it.

gromakovsky commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,109 @@ data EpochSlottingData = EpochSlottingData
+    nextEpochSlottingData = M.singleton 1 esd
+
+-- | Get all epoch index.
+getAllEpochIndex :: SlottingData -> [EpochIndex]
```
Shouldn't it be called `getAllEpochIndices` instead?

ksaric replied:
You are right, I'll rename.

gromakovsky commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,109 @@ data EpochSlottingData = EpochSlottingData
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
+  where
+    decreaseEpochIndex :: EpochIndex -> EpochIndex
```
`EpochIndex` is instance of `Enum`, so this function is `pred`.

ksaric replied:
Never occurred to me, didn't even notice that it's an instance of `Enum`.

gromakovsky commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,109 @@ data EpochSlottingData = EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Compute when the slot started.
+computeSlotStart :: SlotId -> EpochSlottingData -> Timestamp
```
This function should take `LocalSlotIndex` instead of `SlotId`. We can't compute the beginning of arbitrary slot using only one `EpochSlottingData`.

gromakovsky replied:
Also this function is wrong. It should also take `SystemStart`. `epochStartTimeDiff` is what you should add to the system start to get the actual timestamp.

ksaric replied:
Ah, could be. This was one of the things that were changed in `master`. Will fix.

ksaric replied:
It should take `SystemStart` as that was changed in `master`, but I think the rest is good. It takes a `SlotId` which contains the global index position we require to calculate, and the rest we can derive with `SystemStart` and `EpochSlottingData`.

ksaric replied:
![slotting time calculation](https://user-images.githubusercontent.com/6264437/28754626-0f2882da-7549-11e7-8316-b2613626430c.jpg)

ksaric replied:
Also, I think it would be much easier if we added a field `startTime` alongside with `startDiff` in `EpochSlottingData`.
Again, refactoring this would be a breeze if we could "hide" behind a typeclass - https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r130218756

gromakovsky replied:
> Also, I think it would be much easier if we added a field startTime alongside with startDiff in EpochSlottingData.

We can't do it, because we don't want to store any data dependent on the global time in DBs.

ksaric replied:
> We can't do it, because we don't want to store any data dependent on the global time in DBs.

Fair enough, but isn't `EpochSlottingData` `startDiff = currentEpochStart - systemStart` already dependant on the global time? Seems it can't be avoided, no? Otherwise, we can't reconstruct the _timestamp_?

gromakovsky commented on infra/Pos/Slotting/Util.hs:
```diff
@@ -42,57 +42,57 @@ import           Pos.Shutdown           (HasShutdownContext, runIfNotShutdown)
-        esdStartDiff `addTimeDiffToTimestamp`
-        (systemStart + Timestamp (fromIntegral locSlot * convertUnit esdSlotDuration))
+getSlotStartPureFromEpoch :: SlotId -> EpochSlottingData -> Timestamp
+getSlotStartPureFromEpoch = computeSlotStart
```
It seems redundant.

ksaric replied:
Ok, will remove.

gromakovsky commented on infra/Pos/Slotting/Util.hs:
```diff
@@ -42,57 +42,57 @@ import           Pos.Shutdown           (HasShutdownContext, runIfNotShutdown)
 getCurrentSlotFlat = fmap flattenSlotId <$> getCurrentSlot
 
 -- | Get timestamp when given slot starts.
+-- TODO(ks): This function is suspicious. It could be moved to @MonadSlotsData@.
```
As I wrote above, `MonadSlotsData` should be removed.

ksaric replied:
I'll remove the comment.

gromakovsky commented on infra/Pos/Slotting/Util.hs:
```diff
@@ -42,57 +42,57 @@ import           Pos.Shutdown           (HasShutdownContext, runIfNotShutdown)
 -- | Get last known slot duration.
-getLastKnownSlotDuration :: MonadSlotsData m => m Millisecond
-getLastKnownSlotDuration = esdSlotDuration . sdLast <$> getSlottingData
+getLastKnownSlotDuration
```
I think with current renamings it's better to rename it to `getNextEpochSlotDuration` (and update the comment above).

ksaric replied:
I agree, will change.

gromakovsky commented on src/Pos/Generator/Block/Mode.hs:
```diff
@@ -331,10 +343,15 @@ instance MonadBlockGenBase m =>
+    getNextEpochSlottingDataM = getNextEpochSlottingDataDefault
+    getEpochSlottingDataM = getEpochSlottingDataDefault
+    putEpochSlottingDataM = putEpochSlottingDataDefault
+    waitCurrentEpochEqualsM = waitCurrentEpochEqualsDefault
```
Well, we really need to remove `MonadSlotsData`, it's ridiculous 😜

ksaric replied:
:man_shrugging:

pva701 replied:
eto kakaya-to :telega:

ksaric commented on core/Pos/Binary/Class/Instances.hs:
```diff
@@ -243,6 +244,11 @@ instance (Hashable k, Eq k, Bi k, Bi v) => Bi (HM.HashMap k v) where
     put = put . HM.toList
     size = sizeOf HM.toList
 
+instance (Ord k, Eq k, Bi k, Bi v) => Bi (M.Map k v) where
```
Ok.

ksaric commented on infra/Pos/Binary/Infra/Slotting.hs:
```diff
@@ -4,24 +4,22 @@ module Pos.Binary.Infra.Slotting
+instance Bi SlottingData where
+    size = VarSize $ \slottingData -> getSize $ getSlottingDataMap slottingData
+    put slottingData = put $ getSlottingDataMap slottingData
+    get = createSlottingDataUnsafe <$> get
```
I figured that since we can't create an instance outside `Pos.Slotting.Types` it wasn't possible to have it serialized (serialization presupposes creation). Even the `unsafe` constructor has some validation - look at `createSlottingDataUnsafe`.

Other than this and `createInitSlottingData` there isn't a way to construct it and I thought it was deserving a `error` (impossible state reconstructed).

gromakovsky replied:
@ksaric it's true that our code can't create invalid `SlottingData`, but it's not true that nobody else can do it. In general you don't know where this data comes from. It can come from untrusted source.
And using `error` for deserialization errors leads to unpredictable behavior. Long ago we had segfaults because of it.

ksaric replied:
Ok, noted. Will use `fail` before it reaches `error`.

ksaric commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -17,55 +17,79 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
+        -> EpochIndex
+        -> m (Maybe SlotId)
+    findSlot systemStart approxCurTime' epochIndex = do
+        epochSlotData <- getEpochSlottingDataEmpatically epochIndex
```
Ok, will check it out.

ksaric replied:
I just moved the function locally. Removed the extra function, it does simplify things a bit.

ksaric commented:
@gromakovsky Ready for re-re-review.

gromakovsky commented:
I hope to do it today (didn't have time for it yesterday).

gromakovsky commented on update/Pos/Update/Poll/Logic/Base.hs:
```diff
@@ -208,35 +210,39 @@ adoptBlockVersion winningBlk bv = do
+    -> m ()
+updateSlottingData epochIndex = do
     let errFmt =
             ("can't update slotting data, stored penult epoch is "%int%
```
Word `penult` should be removed`.

ksaric replied:
I'm not sure how I missed that, I grepped all over the place.

gromakovsky commented on update/Pos/Update/Poll/Logic/Rollback.hs:
```diff
@@ -43,7 +43,7 @@ rollbackUS (USUndo changedBlockVersions
     whenJust lastAdoptedBV setAdoptedBV
     whenJust prevProposers setEpochProposers
-    whenJust slottingData setSlottingData
+    whenJust slottingData setSlottingData 
```
Redundant space.

gromakovsky replied:
It's still redundant 😔

gromakovsky commented on src/Pos/Block/Slog/Logic.hs:
```diff
@@ -303,8 +307,13 @@ slogRollbackBlocks blunds = do
     sanityCheckDB
     slogPutLastSlots newLastSlots
-    putSlottingData =<< GS.getSlottingData
+    -- TODO(ks): Again, we can use memory, and not break encapsulation.
```
I missed the point here, what's the problem?

ksaric replied:
Nevermind. It's the same as the previous "encapsulation" ramble. Which I still think is valid.

gromakovsky replied:
Is it about `MonadSlotsData` or something different?

ksaric replied:
It's about `MonadSlotsData` and the fact we can use memory to fetch the data, which isn't the case on all these calls.

gromakovsky commented on update/Pos/Update/Poll/Logic/Base.hs:
```diff
@@ -208,35 +210,39 @@ adoptBlockVersion winningBlk bv = do
 
+        let newSlottingData = addEpochSlottingData nextEpochIndex newLast slottingData
+
+        setSlottingData newSlottingData
```
The logic here is correct, but it's quite inefficient. In the previous PR there were these changes https://github.com/input-output-hk/cardano-sl/pull/1107/files#diff-c0fba4e2b0f4c51e5419c9671828d6b5R180
Were did they get lost? 😕

ksaric commented on infra/Pos/Slotting/Error.hs:
```diff
@@ -13,13 +13,18 @@ import           Pos.Core.Types      (SlotId, slotIdF)
 import           Pos.Exception       (cardanoExceptionFromException,
                                       cardanoExceptionToException)
 
--- | Type aggregating run-time errors related to Slotting.
```
:man_shrugging: A mistake, I guess. I'll put it back.

ksaric commented on infra/Pos/Slotting/Impl/Ntp.hs:
```diff
@@ -214,14 +216,14 @@ ntpGetCurrentSlotImpl var = do
 
 ntpGetCurrentSlotBlocking
-    :: NtpMode m
+    :: (NtpMode m, MonadThrow m)
```
Will remove.

ksaric commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+
+-- | Compute when the slot started. We give it @SlotId@, the @SlotId@
+-- @EpochSlottingData@ and find when did that @SlotId@ occur.
+computeSlotStart :: Timestamp -> SlotId -> EpochSlottingData -> Timestamp
```
Ah, now I understand what you meant. Sure, that does make more sense, I'll fix it.

ksaric commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+          -> LocalSlotIndex
+          -> EpochSlottingData
+          -> Timestamp
+      slotTimestamp systemStart' localSlotIndex' epochSlottingData =
```
Ok.

ksaric commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -15,57 +15,104 @@ import           Pos.Core.Slotting           (flattenEpochIndex, unflattenSlotId
+            -> m (Maybe SlotId)
+        iterateIndicesUntilJust indices = do
+          -- Get current index or throw exception if the indices list is empty.
+          currentIndex <- maybeThrow SEUnknownSlot $ head indices
```
Ok, will fix.

ksaric commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -15,57 +15,104 @@ import           Pos.Core.Slotting           (flattenEpochIndex, unflattenSlotId
+        -> Timestamp
+        -> EpochIndex
+        -> m (Maybe SlotId)
+    findSlot systemStart approxCurTime' epochIndex = do
```
Will remove.

ksaric commented on update/Pos/Update/Poll/DBPoll.hs:
```diff
@@ -1,6 +1,6 @@
 {-# LANGUAGE TypeFamilies #-}
 
--- | Instance of MoandPollRead which uses DB.
+-- | Instance of MonadDBPoll which uses DB.
```
Right. And I actually tried to correct it :man_facepalming:

pva701 commented:
@ksaric I will review now, I have to dive into context initially. wait for my review plz

ksaric commented:
@pva701 No problem, I have to fix @gromakovsky review comments anyway. I think I'll do it today.

pva701 commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -15,57 +15,104 @@ import           Pos.Core.Slotting           (flattenEpochIndex, unflattenSlotId
+
+    systemStart  <- getSystemStartM
+    currentIndex <- getCurrentEpochIndexM
+    sdNext       <- getNextEpochSlottingDataM
```
I don't like `getCurrentEpochIndexM` and `getNextEpochSlottingDataM` aren't atomically.
I don't know does it really matter, but can imagine the situation when it matters

ksaric replied:
Good point. Didn't occur to me, but it can happen. I could use a tuple or a structure for returning them both at the same time, and add a comment?
I would like to keep `STM` local in that typeclass, keeps things simple(r).

pva701 replied:
I think we can postpone this until we remove `MonadSlotsData`

pva701 replied:
on the other hand, I don't know when we will do it. so I would prefer tuple, yes.

pva701 commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -15,57 +15,104 @@ import           Pos.Core.Slotting           (flattenEpochIndex, unflattenSlotId
+approxSlotUsingOutdated t = do
+
+    systemStart  <- getSystemStartM
+    currentIndex <- getCurrentEpochIndexM
```
it's minor but I would prefer `currentEpoch` instead of `currentIndex`

ksaric replied:
Ok, will change.

pva701 commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
-    getSystemStart :: m Timestamp
+    getSystemStartM :: m Timestamp
+
+    getAllEpochIndicesM :: m [EpochIndex]
```
I didn't get why this function is here.
AFAIU it's just `[0..nextEpochIndex]`, isn't it?

ksaric replied:
Yes. https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r131144517

gromakovsky replied:
It can be moved outside the class too.

pva701 commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
+
+    getCurrentEpochIndexM :: m EpochIndex
+
+    getNextEpochIndexM :: m EpochIndex
```
Also, I didn't get why it's here, AFAIU it's just `(+1) <$> getCurrentEpochIndexM`

ksaric replied:
Yes, that's true. But I find mine more readable:
- `(+1) <$> getCurrentEpochIndexM`
- `getNextEpochIndexM`

And if we use that all over the code, seems kind of fragile to me.
Using what you proposed we would avoid https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r131131552 but I still prefer to have it written down.

gromakovsky replied:
It can be moved outside the class.

pva701 replied:
my point was: why is it needed in this class? It can be out of the class and just use `getCurrentEpochIndexM`

ksaric replied:
Ok, I guess I can move all the associated functions to `Functions`, now they are mostly in `Types`. Still, typeclass + datatypes seems cleaner than typeclass + datatypes + `Types` + `Functions` + ...
I guess I'm used to looking at the code through OOP :eye:

ksaric replied:
Ok, so what you are saying is that a typeclass must contain only the minimal/crucial amount of information, and we should use other functions we to be derive everything else. Makes sense.

But then you might ask if that information that you have in a typeclass is crucial since you need other functions to redefine it :man_shrugging: 

I think that in this case the code best fits when a typeclass returns `getCurrentNextEpochIndex :: (CurrentEpochIndex, NextEpochIndex)` and we use standard functions like `fst` and `snd` to pick which one we mean.

pva701 commented on infra/Pos/Slotting/MemState/Holder.hs:
```diff
@@ -42,22 +51,72 @@ class HasSlottingVar ctx where
+getNextEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => m EpochSlottingData
+getNextEpochSlottingDataDefault = do
```
tbh I don't like this boilerplate like

```
var <- view slottingVar
atomically $ blah <$> readTVar var
```

ksaric replied:
What boilerplate? :eyes:

pva701 replied:
I meant you used the same pattern of code in these functions. and maybe it makes sense to write a function which will accept getter and do this pattern inside

volhovM replied:
I support pva. You can create local `withSlottingVar` function with the following usage:
```
getNextEpochSlottingDataDefault = 
  withSlottingVar $ \v -> atomically (getNextEpochSlottingData <$> readTVar v)
```

ksaric replied:
Fair enough, I'll refactor.

pva701 commented on infra/Pos/Slotting/MemState/Holder.hs:
```diff
@@ -42,22 +51,72 @@ class HasSlottingVar ctx where
-        penultEpoch <- sdPenultEpoch <$> readTVar var
-        when (penultEpoch < sdPenultEpoch sd) $ writeTVar var sd
+        currentEpoch <- getCurrentEpochIndex <$> readTVar var
+        when (currentEpoch /= target) retry
```
@gromakovsky why did we implement it in this way? if we call w/ `target` less than `currentEpoch` we will get an infinite loop. is it ok? does it make sense to throw an exception ?

gromakovsky replied:
@pva701 this infinite loop won't consume CPU, but I think you are right.
I propose to change this function to wait until current epoch is ≥ target. @ksaric can you do it?

ksaric replied:
Sure, will change.

pva701 commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+-- | Unsafe constructor that can lead to unsafe crash!
+createSlottingDataUnsafe :: Map EpochIndex EpochSlottingData -> SlottingData
+createSlottingDataUnsafe epochSlottingDataMap =
+    if M.size epochSlottingDataMap < 2
```
does it make sense to validate these two epochs are sequential?

ksaric replied:
Yes. Another good point. I'll add that.

pva701 commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+
+-- | Get the current epoch index. Next epoch - 1.
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
```
as I said before `getNextEpochIndex` can be expressed via `getCurrentEpochIndex`  (and vice versa) so I suggest to validate sequential of epochs in the map during putting and remove `getAllEpochIndicesM` and `getNextEpochIndexM` from class. @gromakovsky does it make sense?

pva701 replied:
tbh I don't get why it's map, it looks like a list because the situation, when we put epoch 'x + 1', but didn't put  'x' in this map is strange

ksaric replied:
I'm not sure I'm following, but if it's a list, how are you going to fetch a specific `Epoch` `SlottingData`? Iterate the list - isn't that O(1) vs O(n)?
Again, I might have missed the point.

pva701 replied:
ok, we can use hashmap if we need to have `O(1)` access (actually I am not sure it's really important) but anyway I suggest validating there is the previous epoch in the map when we put new `EpochSlottingData`

gromakovsky replied:
Hm, good point, why not use `Vector`?

pva701 commented on src/Pos/Client/Txp/History.hs:
```diff
@@ -237,15 +236,16 @@ getBlockHistoryDefault
+
+    systemStart <- getSystemStartM
+    bot         <- GS.getBot
+    sd          <- GS.getSlottingData
```
do we still store `SlottingData` in the database?
I thought we will get rid of that datatype.

ksaric replied:
I think @gromakovsky wrote that we have in in both places since we need to pull it into memory at some point if the node get's shut down. Don't really know where that place is :grin:

gromakovsky replied:
@pva701 we definitely need to store `SlottingData` in GState. We can store it using multiple keys (which is easy because this type is a map) or we can store the whole map by one key.
Initially it was implemented by using multiple keys: https://github.com/input-output-hk/cardano-sl/pull/1107/files#diff-495f1c9e499bcfb0abb523efc4bb09b2
Then this logic disappeared for unknown reason (as I asked in https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r130887026).
However, now I think that we probably can store this whole `SlottingData` in one key. We have 1 epoch per 5 days. It's only 71 epoch in a year. One `EpochSlottingData` takes tens of bytes (let's say at most 30, just 2 numbers). So it's only 2 Kb per year. Even in 10 years it will be only 20 Kb which is probably quite cheap (recall that we also store this data in memory, so we rarely read it from DB).

pva701 commented on update/Pos/Update/Poll/Logic/Base.hs:
```diff
@@ -208,35 +210,39 @@ adoptBlockVersion winningBlk bv = do
+    let nextEpochIndex = getNextEpochIndex slottingData
+    let nextEpochSD    = getNextEpochSlottingData slottingData
+
+    if | nextEpochIndex + 1 == epochIndex ->
```
are you sure it's correct?
previously we looked at `penultEpoch` (it means current epoch in new terms)

ksaric replied:
If I do that I get a test failure. 
We receive `EpochSlottingData` for `EpochIndex` that doesn't yet exist - like `nextEpochIndex + 1`, since we do have `nextEpochIndex` - that does make sense, right?

pva701 commented on update/Pos/Update/Poll/PollState.hs:
```diff
@@ -55,7 +55,7 @@ data PollState = PollState
     , _psActivePropsIdx     :: !(HM.HashMap ApplicationName (HashSet UpId))
       -- | Slotting data for this node
-    , _psSlottingData       :: !SlottingData
+    , _psSlottingData       :: !(SlottingData)
```
usefull change :joy:

pva701 commented on test/Test/Pos/Update/PollSpec.hs:
```diff
@@ -71,25 +70,23 @@ data PollAction
     | InsertActiveProposal Poll.ProposalState
     | DeactivateProposal UpId
     | SetSlottingData SlottingData
-    | SetEpochProposers (HashSet StakeholderId)
```
I didn't get why it's removed...

gromakovsky replied:
Yes, that's strange.

volhovM replied:
As far as I see from the PR this particular constructor wasn't used anywhere.

pva701 commented:
There is something mess w/ `SlottingData`, isn't there? I might be mistaken but it seems to be stored in the database and used only in awkward wallet functions :(

gromakovsky commented:
Which mess exactly?

pva701 commented:
>Which mess exactly?
1. where do we use it? (excluding wallet)
2. previously I saw mentions of this datatype in `MonadSlotsData` now it's absence and I thought it's enough to have `HashMap EpochIndex EpochSlottingData` to cover whole our use cases

ksaric commented:
1. It's going to be used in `explorer` also since we need to have block/slot times reconstructed
2. But that is `SlottingData`, just wrapped around that. Probably misunderstood you there.

gromakovsky commented:
> where do we use it? (excluding wallet)

We store in DB and in `SlottingVar`. We need this map to be able to find the time when some old slot started (`getSlotStart`). Previously we were storing this data only for two epochs.

> previously I saw mentions of this datatype in MonadSlotsData now it's absence and I thought it's enough to have HashMap EpochIndex EpochSlottingData to cover whole our use cases

`SlottingData` is `HashMap EpochIndex EpochSlottingData` with size ≥ 2.
But I think we should change it to a `newtype` over `Vector EpochSlottingData`.

volhovM commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
+
+    putEpochSlottingDataM :: EpochIndex -> EpochSlottingData -> m ()
+
+    waitCurrentEpochEqualsM :: EpochIndex -> m ()
```
Ok, so I have nothing logic-related to say except for things already mentioned by pva/gromak. 
Styling comments:
1. Why do you adopt `methodM` pattern here? What's the point?
2. It'd be good to provide at least short documentation to methods here since you've figured out how do they work.
3. Spaces between method declarations are ugly and don't make sense.

ksaric replied:
1. https://github.com/input-output-hk/cardano-sl/pull/1217#issuecomment-318352696 under 3.
2. I agree I can add some comments but I wouldn't say I figured out how they work since there is still a lot of things I'm missing.
3. Reading your comments, you seem to be enjoying very dense code. Which is an aesthetic preference you have every right to. I found it **more readable**, and you can look at the answer here - https://softwareengineering.stackexchange.com/questions/17305/how-do-you-use-blank-lines-in-your-code
Of course, I can return it if you insist, but you might be chasing LOC number more than readability?

volhovM commented on infra/Pos/Slotting/MemState/Class.hs:
```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
+    putEpochSlottingDataM = lift ... putEpochSlottingDataM
+
+    waitCurrentEpochEqualsM = lift . waitCurrentEpochEqualsM
+
```
Oh, an extra line. I hope there are no trailing whitespaces here. Newlines between methods are ugly (as i've already said).

volhovM commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
 instance NFData EpochSlottingData
 
+-- Helpful type aliases
+type CurrentEpochSlottingData = EpochSlottingData
```
I would go with newtypes here.

ksaric replied:
I agree. I would wrap it if it were used outside this module.

volhovM commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+          -- Seems kind of dubious.
+          epochStartTime :: Timestamp
+          epochStartTime = addTimeDiffToTimestamp epochStartTimeDiff systemStart'
+            where
```
Nested `where`s make code less understandable. I think it's better to go with `let` if you don't want to move declaration out of function scope instead of nesting `where`.
In this particular case as `slotTimestamp` has the same arguments as `computeSlotStart` you can move `intSlotIndex` outside to the top level `where` for example.

ksaric replied:
Nesting `where` seems to support the idea of _locality_ while using `let` I can't find an alternative. I find easier to read code with `where` and nested `where` than I do with `let` (especially `let` without type annotations which is common).

volhovM commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+              epochSlotDuration :: Millisecond
+              epochSlotDuration = esdSlotDuration epochSlottingData
+
+
```
😒

ksaric commented on infra/Pos/Slotting/Impl/Sum.hs:
```diff
@@ -40,32 +40,48 @@ data SlottingContextSum
+    )
 
-getCurrentSlotSum :: SlotsSumEnv ctx m => m (Maybe SlotId)
+getCurrentSlotSum
```
Nah, I manually indented that since I find it more readable. I can revert it, but find it's more consistent and _easier to read_ if I have same indentation rule everywhere - not some "if it's 50 spaces wide than this, if it's 60 then...".

ksaric commented:
Looking at the code, it seems to me that `MonadPollRead` should also have/be `MonadSlotsData`. So it should be `class (Monad m, MonadPollRead m, WithLogger m) => MonadPollRead m where ...`
Also, seems `Pos.Update.Poll.DBPoll` is using a redirect as well (if I even understand what a redirect is).
Can I add `MonadSlotsData` to `MonadPollRead`?

Or is the verdict to remove `MonadSlotsData` still the same?

Well, I guess I can try to explain why I don't like the idea of removing `MonadSlotsData` - migrating to `Reader SlottingVar`. I guess people have a lot of work, but I think I have something of value to contribute. I could be wrong :grin:

Let's suppose we remove all functions from `MonadSlotsData` and leave just `getSlottingData` and `setSlottingData` which is something `Reader SlottingVar` will have (effectively).

The functions that currently deal with `SlottingData` which is currently in `Pos.Slotting.Types` in the new PR https://github.com/input-output-hk/cardano-sl/pull/1217 are currently pure. That's just plain wrong and I will fix it.
They should have a `MonadSlotsData` or `MonadSlots` constraint on them since they are being called directly in functions which have `MonadSlotsData`, which is a way to add functionality to a context by using it where the context is being used.
So when we migrate all those functions so they have a `MonadSlotsData` or `MonadSlots` constraint we will be unable to call any function that deals with `MonadSlots` unless we have that constraint in the signature. That means that we will need to add that signature to functions like `updateSlottingData` in `Pos.Update.Poll.Logic.Base`, and others.

So, we effectively have specific functions that are not coupled/organized in any way (like they would be using a typeclass, we would not have to import them individually and we would know what functions are being used from the typeclass signature) that are working with a constraint that is required. And we need to fix all the function constraints from `MonadPoll` to add `MonadSlotsData`/`MonadSlots`/`Reader SlottingVar` in both cases, since that seems like a missing piece.

Let me sum it up. In that case if we use a function, we need to have a `MonadSlotsData` or `MonadSlots` constraint. If we have a constraint, the only way to work with that constraint is to use these functions. 
So what I see is `a -> b & b -> a`. In other words, these functions are the typeclass (belong in the typeclass).

I hope this makes (more?) sense than my previous ramble.

ksaric commented:
@gromakovsky @pva701 @volhovM  I fixed your review comments. I left a `Map` to represent `SlottingData`. Ideally I would use `Stack` but I don't think it's really critical now (`Vector` is an another option as you mentioned @gromakovsky).

I doubt anyone is going to read that monstrosity above :up: so let me sum it up here.

I think we should add `SlotsDefaultEnv` to `MonadPoll...`. We have the exact same thing on `MonadPoll...` but it's using pure functions since it has `getSlottingData` and `putSlottingData` - the thing we want to leave in `HasSlottingVar` - so it's basically the same.

If we go with "pure core, impure shell" approach, then we need to move (pure) functions from `Pos.Slotting.Types` to `Pos.Slotting.MemState.Holder` and remove all extra functions from `Pos.Slotting.MemState.Class` leaving just what you guys mentioned - https://serokell.slack.com/archives/G3K2W68AK/p1502121252517779 which is `type SlotsDefaultEnv ctx m = (MonadReader ctx m, HasSlottingVar ctx, MonadIO m)`?

gromakovsky commented:
> I left a Map to represent SlottingData. Ideally I would use Stack but I don't think it's really critical now (Vector is an another option as you mentioned @gromakovsky).

Ok, `Map` works but is quite confusing, because indices are just sequential numbers. I would like to change it to `Vector`, but after this PR.

> I doubt anyone is going to read that monstrosity above 🆙 so let me sum it up here.

I read it, but to be honest I didn't get most of that message :/

> I think we should add SlotsDefaultEnv to MonadPoll.... We have the exact same thing on MonadPoll... but it's using pure functions since it has getSlottingData and putSlottingData - the thing we want to leave in HasSlottingVar - so it's basically the same.

Adding `MonadSlotsData` to `MonadPoll` is conceptually wrong and error-prone.
1. The reason it's conceptually wrong is the following. `MonadSlotsData` is needed to implement `MonadSlots` whose main goal is to make it possible to know current slot. Contrary, the way `MonadPoll` uses slotting data is that it actually updates it depending on the current blockchain state!
2. The reason it's error-prone is that if we use more than one type class to provide access to data maintained by update system, we need to use more custom implementations.
Look at `Pos.Update.Poll.Trans`. It contains instances of `MonadPollRead` and `MonadPoll` for `PollT` transformer. It stores `PollModifier` internally which contains modifications of underlying data. `MonadPollRead` instance takes this modifications into account.
`MonadPoll` instance doesn't affect actual state, but instead it stores modifications, which are applied later (or not applied if verification fails).
If we use `MonadSlotsData` to operate with slotting data, we will need to also define custom instances for `PollT`. Using `MonadPollRead` and `MonadPoll` for this purpose fits more naturally into the whole picture.
To sum it up:
1. There are two ways to look at slotting data: slotting data to get current slot and slotting data for update system.
2. `MonadSlotsData` provides data to get current slot. There is nothing smart, it's just a reader over `TVar`. It deals with in-memory data.
3. `MonadPollRead` and `MonadPoll` both deal with whole data maintained by update system (read-only access vs writeable access). Slotting data is maintained by update system. There is some smart stuff, specifically `PollT` which uses custom instances. Also there is `RollT`. It's quite convenient to put this data into the same class, instead of using multiple classes.
It deals with global state which is stored in database.

Is the general idea clear? Do you still think we should make `MonadSlotsData` part of `MonadPollRead`?

gromakovsky commented on infra/Pos/Arbitrary/Slotting.hs:
```diff
@@ -2,17 +2,22 @@
+    -- Fixed instance since it's impossible to create and instance
+    -- where one creates @SlottingData@ without at least two parameters.
+    arbitrary = oneof [ createInitSlottingData <$> arbitrary <*> arbitrary ]
     shrink = genericShrink
```
Perhaps `genericShrink` is bad as well, because it can shrink to an invalid value.

ksaric replied:
Removed.

gromakovsky commented on infra/Pos/Slotting/Error.hs:
```diff
@@ -14,12 +14,17 @@ import           Pos.Exception       (cardanoExceptionFromException,
-data SlottingError = SEUnknownSlotStart !SlotId
+data SlottingError
+    = SEUnknownSlotStart !SlotId
+    | SEUnknownSlot
```
It seems this constructor is now unused, so I would delete it.

ksaric replied:
Missed that, will fix.

ksaric replied:
It's used in `getSlotStartEmpatically`.

gromakovsky replied:
`getSlotStartEmpatically` uses `SEUnknownSlotStart`, not `SEUnknownSlot` unless I am blind.

ksaric replied:
You are correct. Blind as a bat :eyeglasses: 
Removed.

gromakovsky commented on infra/Pos/Slotting/Impl/Util.hs:
```diff
@@ -18,54 +18,102 @@ import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp
+
+    systemStart            <- getSystemStartM
+    (currentEpochIndex ,_) <- getCurrentNextEpochIndexM
+    (_, nextSlottingData)  <- getCurrentNextEpochSlottingDataM
```
These two lines will be executed non-atomically which may be a problem.

ksaric replied:
I know, it occurred to me. Seems like the only correct way is to use what you guys proposed - `(MonadReader ctx m, HasSlottingVar ctx)` or similar - a single source of information.
At this point, I think it wouldn't be a problem if I tried to change that, seems like a relatively small change to me. I could be wrong though.

ksaric replied:
And I thought about this, this won't help. It will help with other things and clean some things up nicely, I have an idea how to fix it - instead of `IO` use `STM` and receive all the required data in one go.

gromakovsky commented on infra/Pos/Slotting/MemState/Holder.hs:
```diff
@@ -42,22 +50,80 @@ class HasSlottingVar ctx where
+getEpochSlottingDataDefault ei =
+    withSlottingVar $ lookupEpochSlottingData ei
+
+-- TODO(ks): This is also quite unsafe. We should switch to @addEpochSlottingDataDefault@.
```
I agree. Can you do it?

ksaric replied:
Yes, but then we need to change this too - https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r133762869
The thing I'm concerned with is the rollback and what will happen with the data then.

gromakovsky commented on infra/Pos/Binary/Infra/Slotting.hs:
```diff
@@ -4,24 +4,31 @@ module Pos.Binary.Infra.Slotting
+        -- We don't want to throw a runtime error.
+        checkIfSlottindDataValid slottingDataM = do
+            slottingData <- slottingDataM
+            if M.size slottingData < 2
```
I see you added more checks to `createSlottingDataUnsafe`, but didn't add them here. It makes sense to create a function which creates `SlottingData` and returns `Either Text SlottingData`.

ksaric replied:
You are correct, will add.

gromakovsky commented on src/Pos/Block/Slog/Logic.hs:
```diff
@@ -303,8 +307,16 @@ slogRollbackBlocks blunds = do
+    -- We read from the database and write in the memory.
+    -- TODO(ks): This is unsafe! We don't have control over sequentiality and
+    -- use explicit indexing. It would be better if we have sorted EpochSlotData and
+    -- pass it without the index.
```
Why is it unsafe? `GS.getSlottingData` returns `SlottingData` which has invariant that indices are sequential and start from 0.
If you put all of them into another `SlottingData`, you won't break any invariants.

ksaric replied:
It's generally unsafe because we are using an explicit index.

gromakovsky replied:
Well, if you agree to change `MonadSlotsData` to reader over `TVar SlottingData`, then this code can be simplified by using a single `writeTVar`.

gromakovsky replied:
(and will no longer be unsafe)

ksaric replied:
I don't know. It looks messed up to me. For example, `slogCommon` is called both from _apply_ and _rollback_. 
So in effect no matter what happens, either an _apply_ or _rollback_, it always reads `SlottingData` from database and writes into memory? That seems wrong. What effect will that have when a _rollback_ happens?
If what you said about apply holds (only a single epoch can be computed at a time) and on rollback doesn't, it gets quite complicated.
That's why I don't want to change it more than I have to.

gromakovsky replied:
> For example, `slogCommon` is called both from apply and rollback.

Yes, it's called from apply and rollback, that's why it's called `Common`. It contains functionality which should be used in both functions.

> So in effect no matter what happens, either an apply or rollback, it always reads SlottingData from database and writes into memory? That seems wrong.

Yes, that's what it does. Update system maintains slotting data. It can modify it (`usApplyBlocks`). After it's modified and written into DB we also want to put it into memory. However, there are two problems:
1. The first one was introduced by your PR: in case of rollback it can delete something, not add. I think the best way is to fetch this data from GState (you do it already), and then just use `writeTVar` to put whole `SlottingData` into memory.
2. The second one is that this is fetched before it's actually applied to DB 🤦‍♂️ 
I noticed it only now, but it was present for a long time :/

Also now I think it's more reasonable to have this logic in `usApplyBlocks` (and `usRollbackBlocks`). There we have `PollModifier` which in particular has slotting data modifications. It's ok to add `MonadSlotsData` to `USGlobalApplyMode` (but not to `MonadPoll`).

P. S. I think it can be done after this PR is merged because it took quite a lot of time already.

ksaric commented:
@gromakovsky It's probably more complicated than I can see. What I currently see is `PollT + [PollModifier] = Current state` when the verification passes. And AFAICS, we aren't using `PurePoll`, right? So it's the duality `PollT <-> RollT` where `PollT` contains data for modification and `RollT` for rollback.

I have more comments but I'm sure you see more, so I'll just drop it. I'm still too unfamiliar with a lot of the project.

gromakovsky commented on infra/Pos/Slotting/Types.hs:
```diff
@@ -22,15 +43,152 @@ data EpochSlottingData = EpochSlottingData
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Compute when the slot started. We give it @LocalSlotIndex@, the @LocalSlotIndex@
```
`@LocalSlotIndex@` is duplicated.

gromakovsky commented on test/Test/Pos/Slotting/TypesSpec.hs:
```diff
@@ -0,0 +1,158 @@
+
+----------------------------------------------------------------------------
+-- Test data
+----------------------------------------------------------------------------
```
Usually we write `Spec` before actual tests, but it's not important.

ksaric commented:
@gromakovsky I fixed your review points and simplified `MonadSlotsData` like you guys wanted. At least I hope this is what you wanted to do.
Anyway, there are two issues left:
- https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r134092738
- https://github.com/input-output-hk/cardano-sl/pull/1217/commits/8d295510df3588a4d40e212d94df94ccb72f1d97#diff-5de96c18609f31b47890c0bea47d7f04R99

The first one seems kind of "blinking red lights" to me, or maybe I misunderstood how it works there.
I think the second one was here all along and it might be a cause for some of the bugs (like wallet timestamp bugs?). I will try to fix it tomorrow, will fix it on Monday if I don't make it tomorrow.

gromakovsky commented:
The second one is fixed now, right?

ksaric commented:
> The second one is fixed now, right?

Yes, it's fixed, even though it wasn't as bad as I thought.

ksaric commented:
I added these as issues:
- https://issues.serokell.io/issue/CSL-1528
- https://issues.serokell.io/issue/CSL-1529

When I finish my current tasks (mostly regarding _explorer_) I can switch to these. The first one I can do relatively easy, the second one is more difficult and I would need some (a lot?) feedback before.

ksaric commented:
@gromakovsky I wanted to run it first, but ok :scream:

gromakovsky commented:
Oh, sorry 😬
